### ISSUE TITLE
[Test] sigsuspend() is infinitely blocked/suspended the caller.

### ISSUE THREAD

**Issue by vamsikrishna935** [lsds/sgx-lkl#673]
Original Test Case Description:
     Test Description:
        Verify that sigsuspend() succeeds to change the process's current signal
        mask with the specified signal mask and suspends the process execution
        until the delivery of a signal.
        >Link: https://github.com/linux-test-project/ltp/blob/ed5b99f8832f532c92cb9f0ddaaf3c0bc81368d7/testcases/kernel/syscalls/sigsuspend/sigsuspend01.c#L114

   Test procedure:

     Set a signal handler for SIGALRM.
     Read the current signal mask.
     Set the alarm for 5 seconds.
     Invoke sigsuspend() with an empty signal set.
     The test application expected to come out of the suspended state after receiving the SIGALRM.
     if sigsuspend returned with EINTR then check does signal mask is preserved to the initial value.

    Expected Result:
        sigsuspend() should return after the execution of signal catching
        function and the previous signal mask should be restored.

Problem/Issue:
    sigsuspend() is continuously blocking even after receiving an unblocked signal (SIGALRM). 

     sigsuspend check for the TIF_SIGPENDING flag and if it is enabled it will return else it changes task status to "TASK_INTERRUPTIBLE" and call schedule(). This will move the current task out from the scheduler READY queue.

Link: https://github.com/lsds/lkl/blob/86819ac5446e1fa31ca67db68aa165bbf07b90a9/kernel/signal.c#L4429

GDB Logs:
-----------
```
[Switching to Thread 0x7fff91bef700 (LWP 23542)]

Thread 6 "ENCLAVE" hit Breakpoint 1, 0x00007fe0000c4740 in sigsuspend ()
(gdb) c
Continuing.

Thread 6 "ENCLAVE" hit Breakpoint 1, __do_sys_rt_sigsuspend (sigsetsize=<optimized out>, unewset=<optimized out>) at kernel/signal.c:4458
4458 return sigsuspend(&newset);
(gdb) s
sigsuspend (set=<optimized out>) at kernel/signal.c:4431
4431 current->saved_sigmask = current->blocked;
(gdb) list
4426
4427 #endif
4428
4429 static int sigsuspend(sigset_t *set)
4430 {
4431 current->saved_sigmask = current->blocked;
4432 set_current_blocked(set);
4433
4434 while (!signal_pending(current)) {
4435 __set_current_state(TASK_INTERRUPTIBLE);
(gdb) b kernel/signal.c:4438
Breakpoint 2 at 0x7fe000199f50: kernel/signal.c:4438. (2 locations)
(gdb) c
Continuing.
```
When the SIGALRM is raised, the appropriate task is not woken up while handling the signal by sgx-lkl.
GDB Logs:
```
signal_wake_up_state (t=0x7fe03fe26100, state=1) at kernel/signal.c:764
764 if (!wake_up_state(t, state | TASK_INTERRUPTIBLE))
(gdb) s
wake_up_state (p=0x7fe03fe26100, state=1) at kernel/sched/core.c:2553
2553 {
(gdb) s
2554 return try_to_wake_up(p, state, 0);
(gdb) s
2553 {
(gdb) s
2554 return try_to_wake_up(p, state, 0);
(gdb) s
try_to_wake_up (p=0x7fe03fe26100, state=1, wake_flags=0) at kernel/sched/core.c:2403
2403 if (p == current) {
(gdb) s
2432 raw_spin_lock_irqsave(&p->pi_lock, flags);
(gdb) s
arch_local_irq_save () at ./include/asm-generic/irqflags.h:29
29 flags = arch_local_save_flags();
(gdb) s
arch_local_save_flags () at arch/lkl/kernel/irq.c:166
166 {
(gdb) s
167 return irqs_enabled;
(gdb) s
166 {
(gdb) s
168 }
(gdb) s
arch_local_irq_save () at ./include/asm-generic/irqflags.h:30
30 arch_local_irq_restore(ARCH_IRQ_DISABLED);
(gdb) s
29 flags = arch_local_save_flags();
(gdb) s
30 arch_local_irq_restore(ARCH_IRQ_DISABLED);
(gdb) s
arch_local_irq_restore (flags=0) at arch/lkl/kernel/irq.c:172
172 if (flags == ARCH_IRQ_ENABLED && irqs_enabled == ARCH_IRQ_DISABLED &&
(gdb) s
175 irqs_enabled = flags;
(gdb) s
try_to_wake_up (p=0x7fe03fe26100, state=<optimized out>, wake_flags=0) at kernel/sched/core.c:2434
2434 if (!(p->state & state))
(gdb) s
2400 int cpu, success = 0;
(gdb) s
2434 if (!(p->state & state))
(gdb) s
2526 raw_spin_unlock_irqrestore(&p->pi_lock, flags);
(gdb) s
arch_local_irq_restore (flags=0) at arch/lkl/kernel/irq.c:172
172 if (flags == ARCH_IRQ_ENABLED && irqs_enabled == ARCH_IRQ_DISABLED &&
(gdb) s
175 irqs_enabled = flags;
(gdb) s
try_to_wake_up (p=0x0, state=1, wake_flags=0) at kernel/sched/core.c:2533
2533 }
(gdb) s
wake_up_state (p=<optimized out>, state=<optimized out>) at kernel/sched/core.c:2555
2555 }
(gdb) s
signal_wake_up_state (t=0x7fe03fe26100, state=1) at kernel/signal.c:766
766 }
```
Complete Logs for Reference:
```
[    0.102769] EXT4-fs (vda): mounted filesystem with ordered data mode. Opts: 

[Switching to Thread 0x7fff91bef700 (LWP 22983)]

Thread 6 "ENCLAVE" hit Breakpoint 1, 0x00007fe0000c4740 in sigsuspend ()
(gdb) b signal_wake_up_state
Breakpoint 2 at 0x7fe000196460: file kernel/signal.c, line 755.
(gdb) c
Continuing.

Thread 6 "ENCLAVE" hit Breakpoint 1, __do_sys_rt_sigsuspend (sigsetsize=<optimized out>, unewset=<optimized out>) at kernel/signal.c:4458
4458		return sigsuspend(&newset);
(gdb) s
sigsuspend (set=<optimized out>) at kernel/signal.c:4431
4431		current->saved_sigmask = current->blocked;
(gdb) s
4432		set_current_blocked(set);
(gdb) n
(gdb) s
4432		set_current_blocked(set);
(gdb) s
set_current_blocked (newset=<optimized out>) at kernel/signal.c:4432
4432		set_current_blocked(set);
(gdb) s
sigdelsetmask (mask=<optimized out>, set=<optimized out>) at ./include/linux/signal.h:217
217		set->sig[0] &= ~mask;
(gdb) s
sigsuspend (set=<optimized out>) at kernel/signal.c:4431
4431		current->saved_sigmask = current->blocked;
(gdb) s
4432		set_current_blocked(set);
(gdb) s
set_current_blocked (newset=<optimized out>) at kernel/signal.c:2900
2900		__set_current_blocked(newset);
(gdb) s
__set_current_blocked (newset=0x7fe040a74e98) at kernel/signal.c:2905
2905		struct task_struct *tsk = current;
(gdb) s
2904	{
(gdb) s
__set_current_blocked (newset=0x7fe040a74e98) at kernel/signal.c:2905
2905		struct task_struct *tsk = current;
(gdb) s
2911		if (sigequalsets(&tsk->blocked, newset))
(gdb) s
2917	}
(gdb) s
sigsuspend (set=<optimized out>) at kernel/signal.c:4434
4434		while (!signal_pending(current)) {
(gdb) s
current_thread_info () at ./arch/lkl/include/asm/thread_info.h:46
46		return _current_thread_info;
(gdb) s
sigsuspend (set=<optimized out>) at kernel/signal.c:4434
4434		while (!signal_pending(current)) {
(gdb) s
signal_pending (p=<optimized out>) at kernel/signal.c:4434
4434		while (!signal_pending(current)) {
(gdb) s
test_tsk_thread_flag (flag=<optimized out>, tsk=<optimized out>) at ./include/linux/sched.h:1746
1746		return test_ti_thread_flag(task_thread_info(tsk), flag);
(gdb) s
test_ti_thread_flag (flag=<optimized out>, ti=<optimized out>) at ./include/linux/sched.h:1746
1746		return test_ti_thread_flag(task_thread_info(tsk), flag);
(gdb) s
test_bit (addr=<optimized out>, nr=<optimized out>) at ./include/asm-generic/bitops/non-atomic.h:106
106		return 1UL & (addr[BIT_WORD(nr)] >> (nr & (BITS_PER_LONG-1)));
(gdb) s
sigsuspend (set=<optimized out>) at kernel/signal.c:4434
4434		while (!signal_pending(current)) {
(gdb) s
4435			__set_current_state(TASK_INTERRUPTIBLE);
(gdb) s
4436			schedule();
(gdb) s
schedule () at kernel/sched/core.c:3942
3942		struct task_struct *tsk = current;
(gdb) c
Continuing.

Thread 6 "ENCLAVE" hit Breakpoint 2, signal_wake_up_state (t=0x7fe03fe26100, state=0) at kernel/signal.c:755
755	{
(gdb) s
756		set_tsk_thread_flag(t, TIF_SIGPENDING);
(gdb) s
set_tsk_thread_flag (flag=<optimized out>, tsk=<optimized out>) at kernel/signal.c:756
756		set_tsk_thread_flag(t, TIF_SIGPENDING);
(gdb) s
set_ti_thread_flag (flag=<optimized out>, ti=<optimized out>) at ./include/linux/thread_info.h:55
55		set_bit(flag, (unsigned long *)&ti->flags);
(gdb) s
signal_wake_up_state (t=0x7fe03fe26100, state=0) at kernel/signal.c:755
755	{
(gdb) s
764		if (!wake_up_state(t, state | TASK_INTERRUPTIBLE))
(gdb) s
756		set_tsk_thread_flag(t, TIF_SIGPENDING);
(gdb) s
set_tsk_thread_flag (flag=<optimized out>, tsk=<optimized out>) at kernel/signal.c:756
756		set_tsk_thread_flag(t, TIF_SIGPENDING);
(gdb) s
set_ti_thread_flag (flag=<optimized out>, ti=<optimized out>) at kernel/signal.c:756
756		set_tsk_thread_flag(t, TIF_SIGPENDING);
(gdb) s
set_bit (p=<optimized out>, nr=<optimized out>) at kernel/signal.c:756
756		set_tsk_thread_flag(t, TIF_SIGPENDING);
(gdb) s
atomic_long_or (v=<optimized out>, i=<optimized out>) at ./include/asm-generic/atomic-long.h:328
328		atomic64_or(i, v);
(gdb) s
set_ti_thread_flag (flag=<optimized out>, ti=<optimized out>) at ./include/linux/thread_info.h:55
55		set_bit(flag, (unsigned long *)&ti->flags);
(gdb) s
set_bit (p=<optimized out>, nr=<optimized out>) at ./include/linux/thread_info.h:55
55		set_bit(flag, (unsigned long *)&ti->flags);
(gdb) s
atomic_long_or (v=<optimized out>, i=<optimized out>) at ./include/asm-generic/atomic-long.h:328
328		atomic64_or(i, v);
(gdb) s
atomic64_or (a=4, v=0x7fe03f9da008) at lib/atomic64.c:125
125	ATOMIC64_OPS(or, |=)
(gdb) s
arch_local_irq_save () at ./include/asm-generic/irqflags.h:29
29		flags = arch_local_save_flags();
(gdb) s
arch_local_save_flags () at arch/lkl/kernel/irq.c:166
166	{
(gdb) s
167		return irqs_enabled;
(gdb) s
166	{
(gdb) s
168	}
(gdb) s
arch_local_irq_save () at ./include/asm-generic/irqflags.h:30
30		arch_local_irq_restore(ARCH_IRQ_DISABLED);
(gdb) s
29		flags = arch_local_save_flags();
(gdb) s
30		arch_local_irq_restore(ARCH_IRQ_DISABLED);
(gdb) s
arch_local_irq_restore (flags=0) at arch/lkl/kernel/irq.c:172
172		if (flags == ARCH_IRQ_ENABLED && irqs_enabled == ARCH_IRQ_DISABLED &&
(gdb) s
175		irqs_enabled = flags;
(gdb) s
atomic64_or (a=4, v=0x7fe03f9da008) at lib/atomic64.c:125
125	ATOMIC64_OPS(or, |=)
(gdb) s
arch_local_irq_restore (flags=0) at arch/lkl/kernel/irq.c:172
172		if (flags == ARCH_IRQ_ENABLED && irqs_enabled == ARCH_IRQ_DISABLED &&
(gdb) s
175		irqs_enabled = flags;
(gdb) s
signal_wake_up_state (t=0x7fe03fe26100, state=1) at kernel/signal.c:764
764		if (!wake_up_state(t, state | TASK_INTERRUPTIBLE))
(gdb) s
wake_up_state (p=0x7fe03fe26100, state=1) at kernel/sched/core.c:2553
2553	{
(gdb) s
2554		return try_to_wake_up(p, state, 0);
(gdb) s
2553	{
(gdb) s
2554		return try_to_wake_up(p, state, 0);
(gdb) s
try_to_wake_up (p=0x7fe03fe26100, state=1, wake_flags=0) at kernel/sched/core.c:2403
2403		if (p == current) {
(gdb) s
2432		raw_spin_lock_irqsave(&p->pi_lock, flags);
(gdb) s
arch_local_irq_save () at ./include/asm-generic/irqflags.h:29
29		flags = arch_local_save_flags();
(gdb) s
arch_local_save_flags () at arch/lkl/kernel/irq.c:166
166	{
(gdb) s
167		return irqs_enabled;
(gdb) s
166	{
(gdb) s
168	}
(gdb) s
arch_local_irq_save () at ./include/asm-generic/irqflags.h:30
30		arch_local_irq_restore(ARCH_IRQ_DISABLED);
(gdb) s
29		flags = arch_local_save_flags();
(gdb) s
30		arch_local_irq_restore(ARCH_IRQ_DISABLED);
(gdb) s
arch_local_irq_restore (flags=0) at arch/lkl/kernel/irq.c:172
172		if (flags == ARCH_IRQ_ENABLED && irqs_enabled == ARCH_IRQ_DISABLED &&
(gdb) s
175		irqs_enabled = flags;
(gdb) s
try_to_wake_up (p=0x7fe03fe26100, state=<optimized out>, wake_flags=0) at kernel/sched/core.c:2434
2434		if (!(p->state & state))
(gdb) s
2400		int cpu, success = 0;
(gdb) s
2434		if (!(p->state & state))
(gdb) s
2526		raw_spin_unlock_irqrestore(&p->pi_lock, flags);
(gdb) s
arch_local_irq_restore (flags=0) at arch/lkl/kernel/irq.c:172
172		if (flags == ARCH_IRQ_ENABLED && irqs_enabled == ARCH_IRQ_DISABLED &&
(gdb) s
175		irqs_enabled = flags;
(gdb) s
try_to_wake_up (p=0x0, state=1, wake_flags=0) at kernel/sched/core.c:2533
2533	}
(gdb) s
wake_up_state (p=<optimized out>, state=<optimized out>) at kernel/sched/core.c:2555
2555	}
(gdb) s
signal_wake_up_state (t=0x7fe03fe26100, state=1) at kernel/signal.c:766
766	}
(gdb) s
complete_signal (sig=14, p=0x7fe03fe26100, type=PIDTYPE_PID) at kernel/signal.c:1058
1058	}
(gdb) s
__send_signal (sig=14, info=0x1, t=0x7fe03fe26100, type=PIDTYPE_TGID, force=<optimized out>) at kernel/signal.c:1071
1071		int ret = 0, result;
(gdb) s
1179	}
(gdb) s
send_signal (sig=14, info=0x1, t=0x0, type=PIDTYPE_TGID) at kernel/signal.c:1237
1237	}
(gdb) s
do_send_sig_info (sig=14, info=0x1, p=0x7fe03fe26100, type=PIDTYPE_TGID) at kernel/signal.c:1286
1286			unlock_task_sighand(p, &flags);
(gdb) s
unlock_task_sighand (flags=<optimized out>, task=<optimized out>) at kernel/signal.c:1286
1286			unlock_task_sighand(p, &flags);
(gdb) bt
#0  unlock_task_sighand (flags=<optimized out>, task=<optimized out>) at kernel/signal.c:1286
#1  do_send_sig_info (sig=14, info=0x1, p=0x7fe03fe26100, type=PIDTYPE_TGID) at kernel/signal.c:1286
#2  0x00007fe000198cc3 in group_send_sig_info (sig=1071800576, info=0xe, p=0x1, type=PIDTYPE_TGID) at kernel/signal.c:1408
#3  0x00007fe000198dda in kill_pid_info (sig=0, info=0x1, pid=0x0) at kernel/signal.c:1442
#4  0x00007fe0001d0f3a in it_real_fn (timer=<optimized out>) at kernel/time/itimer.c:143
#5  0x00007fe0001c6a4a in __run_hrtimer (cpu_base=<optimized out>, now=<optimized out>, flags=<optimized out>, timer=<optimized out>, 
    base=<optimized out>) at kernel/time/hrtimer.c:1389
#6  __hrtimer_run_queues (now=<optimized out>, flags=1, active_mask=<optimized out>, cpu_base=<optimized out>) at kernel/time/hrtimer.c:1451
#7  0x00007fe0001c728c in hrtimer_interrupt (dev=<optimized out>) at kernel/time/hrtimer.c:1509
#8  0x00007fe000188669 in timer_irq_handler (irq=<optimized out>, dev_id=<optimized out>) at arch/lkl/kernel/time.c:96
#9  0x00007fe0001bc96a in __handle_irq_event_percpu (desc=<optimized out>, flags=0x1) at kernel/irq/handle.c:149
#10 0x00007fe0001bca5e in handle_irq_event_percpu (desc=0x7fe03fe26100) at kernel/irq/handle.c:189
#11 0x00007fe0001bcab2 in handle_irq_event (desc=0x7fe03fe26100) at kernel/irq/handle.c:206
#12 0x00007fe0001c0230 in handle_simple_irq (desc=0x0) at kernel/irq/chip.c:553
#13 0x00007fe0001bc224 in generic_handle_irq_desc (desc=<optimized out>) at ./include/linux/irqdesc.h:156
#14 generic_handle_irq (irq=<optimized out>) at kernel/irq/irqdesc.c:644
#15 0x00007fe040874e40 in ?? ()
#16 0x00007fe0001884d0 in run_irq (irq=1) at arch/lkl/kernel/irq.c:65
#17 0x0000000000000001 in ?? ()
#18 0x0000000000000001 in ?? ()
#19 0x0000000000000002 in ?? ()
#20 0x00007fe040874e80 in ?? ()
#21 0x00007fe000188316 in deliver_irq (index=<optimized out>, bit=<optimized out>) at arch/lkl/kernel/irq.c:119
#22 for_each_bit (j=<optimized out>, f=<optimized out>, word=140601121202432) at arch/lkl/kernel/irq.c:111
#23 check_irq_status (unused=<optimized out>, i=<optimized out>) at arch/lkl/kernel/irq.c:124
#24 for_each_bit (j=<optimized out>, f=<optimized out>, word=14) at arch/lkl/kernel/irq.c:111
#25 run_irqs () at arch/lkl/kernel/irq.c:129
#26 0x00007fe040874e70 in ?? ()
#27 0x00007fe03f9ea000 in ?? ()
#28 0x00007fe03f9ea000 in ?? ()
#29 0x00007fe03fc1a100 in ?? ()
#30 0x0000000000000001 in ?? ()
#31 0x00007fe03fe25b80 in ?? ()
#32 0x00007fe040874ea0 in ?? ()
#33 0x00007fe000188473 in arch_local_irq_restore (flags=<optimized out>) at arch/lkl/kernel/irq.c:174
#34 0x00007fe0008a36c0 in hrtimer_bases ()
#35 0x0000000000000001 in ?? ()
#36 0x00007fe040874ee0 in ?? ()
#37 0x00007fe0001aa8e6 in arch_local_irq_enable () at ./include/asm-generic/irqflags.h:47
#38 finish_lock_switch (rq=<optimized out>) at kernel/sched/core.c:3004
---Type <return> to continue, or q <return> to quit---
#39 finish_task_switch (prev=0x0) at kernel/sched/core.c:3104
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb) s
spin_unlock_irqrestore (flags=<optimized out>, lock=<optimized out>) at ./include/linux/spinlock.h:393
393		raw_spin_unlock_irqrestore(&lock->rlock, flags);
(gdb) s
do_send_sig_info (sig=14, info=0x1, p=0x7fe03fe26100, type=PIDTYPE_TGID) at kernel/signal.c:1285
1285			ret = send_signal(sig, info, p, type);
(gdb) s
1286			unlock_task_sighand(p, &flags);
(gdb) s
unlock_task_sighand (flags=<optimized out>, task=<optimized out>) at kernel/signal.c:1286
1286			unlock_task_sighand(p, &flags);
(gdb) s
spin_unlock_irqrestore (flags=<optimized out>, lock=<optimized out>) at ./include/linux/spinlock.h:393
393		raw_spin_unlock_irqrestore(&lock->rlock, flags);
(gdb) s
arch_local_irq_restore (flags=0) at arch/lkl/kernel/irq.c:172
172		if (flags == ARCH_IRQ_ENABLED && irqs_enabled == ARCH_IRQ_DISABLED &&
(gdb) s
175		irqs_enabled = flags;
(gdb) c
Continuing.
q
q

^C
Thread 1 "sgx-lkl-run-oe" received signal SIGINT, Interrupt.
[Switching to Thread 0x7ffff7fea300 (LWP 22974)]
0x00007ffff72e7d2d in __GI___pthread_timedjoin_ex (threadid=140735638599424, thread_return=0x7fffffffd120, abstime=0x0, block=<optimized out>)
    at pthread_join_common.c:89
89	pthread_join_common.c: No such file or directory.
(gdb) q
```

**Comment by hukoyu** [comment]
cc @davidchisnall @prp @paulcallen @mikbras

**Comment by SeanTAllen** [comment]
@vamsikrishna935 this is probably https://github.com/lsds/sgx-lkl/issues/209

**Comment by prp** [comment]
@SeanTAllen @vamsikrishna935 I think it's rather https://github.com/lsds/sgx-lkl/issues/644.

**Comment by SeanTAllen** [comment]
@prp maybe both. :wink:

**Comment by prp** [comment]
I think that the lack of proper stack frames (as mentioned in #644) is the underlying cause of the problems in #209.

**Comment by SeanTAllen** [comment]
@prp I think you are probably right.

**Comment by davidchisnall** [comment]
Duplicate of #209
